### PR TITLE
TF-TRT implement Range and RangeOptimal optimization profile strategies

### PR DESCRIPTION
**Pull request by tfeher**
This PR implements Range and RangeOptimal optimization profile strategies for TF-TRT dynamic shape mode.

Tagging @bixia1 for review and @DEKHTIARJonathan for visibility.

Tracker: #45481

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 1]
```diff
@@ -78,6 +78,52 @@ void TrtShapeOptimizationProfile::ImplicitBatchModeCompatibleStrategy(
   }
 }
 
+// Create a new Dims vector where the dimensions are the result of a binary
+// operation between dimensions in lhs and rhs.
+template <typename BinaryOperation>
+Status ShapeProfileBinaryOp(const std::vector<nvinfer1::Dims>& lhs,
+                            const std::vector<nvinfer1::Dims>& rhs,
+                            BinaryOperation op,
+                            std::vector<nvinfer1::Dims>* out) {
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc:87]: Suggest to make this change:
// For each dimension d, accumulates the dimension value in rhs[d] to lhs[d] to construct shape for shape profiling.
// The BinaryOperation may be std::min, std::max etc.
template <typename BinaryOperation>
Status ShapeProfileBinaryOp(std::vector<nvinfer1::Dims>* lhs,
                            const std::vector<nvinfer1::Dims>& rhs,
                            BinaryOperation op) {

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 2]
```diff
@@ -78,6 +78,52 @@ void TrtShapeOptimizationProfile::ImplicitBatchModeCompatibleStrategy(
...
+    const std::vector<std::vector<nvinfer1::Dims>>& collected_shapes) {
+  if (collected_shapes.empty()) return Status::OK();
+
+  std::vector<nvinfer1::Dims> min = collected_shapes[0];
+  std::vector<nvinfer1::Dims> max = min;
+
+  for (int i = 1; i < collected_shapes.size(); i++) {
+    TF_RETURN_IF_ERROR(ShapeProfileBinaryOp(
+        min, collected_shapes[i], [](int a, int b) { return std::min(a, b); },
+        &min));
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc:115]: Can we replace the lambda with std::min?

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc [hunk 3]
```diff
@@ -125,9 +154,19 @@ class TrtShapeOptimizationProfileTest : public ::testing::Test {
...
 
-TEST_F(TrtShapeOptimizationProfileTest, Static) {
+INSTANTIATE_TEST_CASE_P(
+    OptProfilesTestInstantiation, TrtShapeOptimizationProfileTest,
+    ::testing::Values(ProfileStrategy::kRange, ProfileStrategy::kOptimal,
+                      ProfileStrategy::kRangeOptimal,
+                      ProfileStrategy::kImplicitBatchModeCompatible));
+
+TEST_P(TrtShapeOptimizationProfileTest, Static) {
+  // Static mode does not depend on strategies, we only once.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:167]: Sorry that I don't understand this comment. What is "we only once"?

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc (hunk 2 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: Not simply, it is overleaded and the lambda is used to resolve ambiguity. Alternatively, we could [cast it to a function pointer](https://stackoverflow.com/a/47109361/9018949) if you prefer that solution.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc (hunk 1 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc]: I have changed the function to return the results in the first arg, and improved the docstring.

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc (hunk 3 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc]: Hmm, the verb got deleted from the sentence. Fixed it.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the comments, I have addressed the issues.

**Review by bixia1 — APPROVED** [review]

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc [hunk 4]
```diff
@@ -110,6 +112,33 @@ class TrtShapeOptimizationProfileTest : public ::testing::Test {
...
   }
 
+  void CheckProfile(const std::vector<nvinfer1::Dims3>& dimvec,
+                    TrtShapeOptimizationProfile* profile, bool has_prof,
+                    bool test_optimality) {
+    std::vector<TensorShape> shape_vec = DimVecToShapeVec(dimvec);
+    int idx = profile->GetProfileNumber(shape_vec);
+    ASSERT_EQ(idx >= 0, has_prof);
+    if (idx < 0) return;
+    int prof_idx = exec_contexts_[idx]->getOptimizationProfile();
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:122]: This API is not available in TRT 5.
There are other similar errors when running the test with TRT 5:

third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:122:41: error: no member named 'getOptimizationProfile' in 'nvinfer1::IExecutionContext'
    int prof_idx = exec_contexts_[idx]->getOptimizationProfile();
                   ~~~~~~~~~~~~~~~~~~~~~^
third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:126:34: error: no member named 'OptProfileSelector' in namespace 'nvinfer1'
          j, prof_idx, nvinfer1::OptProfileSelector::kMIN);
                       ~~~~~~~~~~^
third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:125:36: error: no member named 'getProfileDimensions' in 'nvinfer1::ICudaEngine'; did you mean 'getBindingDimensions'?
      nvinfer1::Dims min = engine->getProfileDimensions(
                                   ^~~~~~~~~~~~~~~~~~~~
                                   getBindingDimensions
blaze-out/k8-cuda101-opt/bin/third_party/tensorrt/_virtual_includes/nv_infer_static_cuda_10_1/third_party/tensorrt/NvInfer.h:4985:18: note: 'getBindingDimensions' declared here
    virtual Dims getBindingDimensions(int bindingIndex) const = 0;
                 ^
third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:128:34: error: no member named 'OptProfileSelector' in namespace 'nvinfer1'
          j, prof_idx, nvinfer1::OptProfileSelector::kMAX);
                       ~~~~~~~~~~^
third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:127:36: error: no member named 'getProfileDimensions' in 'nvinfer1::ICudaEngine'; did you mean 'getBindingDimensions'?
      nvinfer1::Dims max = engine->getProfileDimensions(
                                   ^~~~~~~~~~~~~~~~~~~~
                                   getBindingDimensions
blaze-out/k8-cuda101-opt/bin/third_party/tensorrt/_virtual_includes/nv_infer_static_cuda_10_1/third_party/tensorrt/NvInfer.h:4985:18: note: 'getBindingDimensions' declared here
    virtual Dims getBindingDimensions(int bindingIndex) const = 0;
                 ^
third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:130:34: error: no member named 'OptProfileSelector' in namespace 'nvinfer1'
          j, prof_idx, nvinfer1::OptProfileSelector::kOPT);
                       ~~~~~~~~~~^
third_party/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:129:36: error: no member named 'getProfileDimensions' in 'nvinfer1::ICudaEngine'; did you mean 'getBindingDimensions'?
      nvinfer1::Dims opt = engine->getProfileDimensions(
                                   ^~~~~~~~~~~~~~~~~~~~
                                   getBindingDimensions
blaze-out/k8-cuda101-opt/bin/third_party/tensorrt/_virtual_includes/nv_infer_static_cuda_10_1/third_party/tensorrt/NvInfer.h:4985:18: note: 'getBindingDimensions' declared here
    virtual Dims getBindingDimensions(int bindingIndex) const = 0;

**Review by bixia1 — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@tfeher  Can you please check @bixia1's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@tfeher Any update on this PR? Please. Thanks!

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc (hunk 4 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles_test.cc:123]: I have added an ifdef to guard the TRT6+ API calls.

**Review by tfeher — COMMENTED** [review]
Sorry for the delay, I have updated the PR to fix the issue with TRT5 compilation.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**